warehouse/legacy/api/json.py

```diff
@@ -217,3 +217,120 @@ def json_release_slash(release, request):
         ),
         headers=_CORS_HEADERS,
     )
+
+
+@view_config(
+    route_name="legacy.api.json.latest",
+    context=Project,
+    renderer="json",
+    decorator=_CACHE_DECORATOR,
+)
+def json_latest(project, request):
+    release = project.latest_version
+    version = release.version if release else None
+
+    if version is None:
+        return HTTPNotFound(headers=_CORS_HEADERS)
+
+    release = (
+        request.db.query(Release)
+        .filter(Release.project == project)
+        .filter(Release.version == version)
+        .first()
```

Comment:
Is `.first()` the best way to trim this to one result? Would it be faster to use, e.g., `.one()` or `.limit(1)`?

Reply:
`one` would raise if there are more than one result, which is not what you want.
`limit()` would return an iterable which is not what you want. There's no reason calling SQLAlchemy's function exactly for its main use-case would be suboptimal. `first` is absolutely fine, as far as I can tell.

Here's the implementation for first:
https://github.com/sqlalchemy/sqlalchemy/blob/201893e3f6f7f3bdc8b9c93cd25092cd6fa8cba3/lib/sqlalchemy/engine/result.py#L1476-L1487
Here's the implementation for one:
https://github.com/sqlalchemy/sqlalchemy/blob/201893e3f6f7f3bdc8b9c93cd25092cd6fa8cba3/lib/sqlalchemy/engine/result.py#L1501-L1511

Reply:
In this case, if there was 0 or multiple results, it would be a bug, so it could make sense to call `one()` (if we had mypy, it would complain that release might be None here). `limit(1)` wouldn't bring a lot to the table since you don't want an iterable. I think it makes the most sense to use the function whose Python API is most adapted to what you want and trust sqlalchemy to optimize it correctly.

Reply:
Nod, I like that change; I'd looked at the docs but hadn't appreciated the raising behavior of `.one()`.